controllers/argocd/multi-cluster-controller.go

```diff
@@ -0,0 +1,223 @@
+*/
+
+package argocd
+
+import (
+	"context"
+	"encoding/base64"
+	"encoding/json"
+	"github.com/go-logr/logr"
+	v1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/record"
+	"kubesphere.io/devops/pkg/utils/k8sutil"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+)
+
+//+kubebuilder:rbac:groups=cluster.kubesphere.io,resources=clusters,verbs=get;list;watch
+//+kubebuilder:rbac:groups="",resources=secrets,verbs=get;list;create;update
+
+// MultiClusterReconciler represents a controller to sync cluster to ArgoCD cluster
+type MultiClusterReconciler struct {
+	client.Client
+	log      logr.Logger
+	recorder record.EventRecorder
+
+	argocdNamespace string
+}
+
+// Reconcile is the entrypoint of the controller
+func (r *MultiClusterReconciler) Reconcile(req ctrl.Request) (result ctrl.Result, err error) {
+	var cluster *unstructured.Unstructured
+	if cluster, err = getCluster(r.Client, req.NamespacedName); err != nil {
+		err = client.IgnoreNotFound(err)
+		return
+	}
+
+	// find the namespace of ArgoCD, and cache it
+	if r.argocdNamespace == "" {
+		if r.argocdNamespace = r.findArgoCDNamespace(); r.argocdNamespace == "" {
+			r.log.V(6).Info("cannot found the namespace of ArgoCD")
+			return
+		}
+	}
+
+	// ignore the host cluster
+	if ignore(cluster) {
+		return
+	}
+
+	ownerRef := getOwnerReference(cluster)
+	argoCluster := createArgoCluster(cluster)
+	argoCluster.SetNamespace(r.argocdNamespace)
+	err = r.updateOrCreate(argoCluster, ownerRef)
+	return
+}
+
+func getOwnerReference(object *unstructured.Unstructured) (ref metav1.OwnerReference) {
+	ref = metav1.OwnerReference{
+		APIVersion: object.GetAPIVersion(),
+		Kind:       object.GetKind(),
+		Name:       object.GetName(),
+		UID:        object.GetUID(),
+	}
+	return
+}
+
+func (r *MultiClusterReconciler) updateOrCreate(cluster *v1.Secret, ownerRef metav1.OwnerReference) (err error) {
+	ctx := context.Background()
+	existingArgoCluster := cluster.DeepCopy()
+	if err = r.Get(ctx, types.NamespacedName{
+		Namespace: cluster.Namespace,
+		Name:      cluster.Name,
+	}, existingArgoCluster); err != nil {
+		if apierrors.IsNotFound(err) {
+			// create the argo cluster
+			k8sutil.SetOwnerReference(cluster, ownerRef)
+			err = r.Create(ctx, cluster)
+		}
+		return
+	}
+	existingArgoCluster.Data = cluster.Data
+	k8sutil.SetOwnerReference(existingArgoCluster, ownerRef)
+	err = r.Update(ctx, existingArgoCluster)
+	return
+}
+
+// let's assume there is only one ArgoCD existing
+// TODO consider to have a better solution to figure the namespace of ArgoCD
+func (r *MultiClusterReconciler) findArgoCDNamespace() string {
+	secretList := &v1.SecretList{}
+	if err := r.List(context.Background(), secretList); err == nil {
```

Comment:
It's better to use `r.Get(...)` to find `argocd-secret` Secret instead of listing and foreach. Please correct me if I missed something.

Reply:
The problem is that I don't know the secret namespace. So it looks like no way to use `Get`.